File: tools/devserver/handler.py
# minimoni - zero-dependency system monitoring
# Copyright (C) 2026 Javier Beaumont <javierbeaumont@users.noreply.github.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

"""
HTTP request handler for the minimoni dev server.

A handler is bound to an AppState (version, real config fields, temp_max,
scenario) via make_handler(), so there is no module-global state. Raw mock
metrics are converted to the configured units by units.py.
"""

from dataclasses import dataclass
from http.server import BaseHTTPRequestHandler
from json import dumps
from logging import getLogger
from os import sep
from os.path import abspath, basename, dirname, isfile, join, realpath
from time import sleep
from urllib.parse import parse_qs, urlparse

from mock_data import JSON, clamp_points, current_snapshot, make_points
from units import to_current, to_point

log = getLogger("minimoni-dev")

REPO_ROOT = join(dirname(abspath(__file__)), "..", "..")
DASHBOARD_DIR = join(REPO_ROOT, "dashboard")
DASHBOARD = join(DASHBOARD_DIR, "index.html")

STATIC_FILES: dict[str, str] = {
    "app.js": "text/javascript",
    "style.css": "text/css",
    "favicon.svg": "image/svg+xml",
}


@dataclass
class AppState:
    version: str
    config_fields: JSON
    temp_max: float
    scenario: str


def make_handler(state: AppState) -> type[BaseHTTPRequestHandler]:
    """Build a request handler bound to the given app state."""

    def current() -> JSON:
        # converted mocked metrics + version + real config fields
        raw = current_snapshot(state.scenario)
        return {
            **to_current(raw, state.config_fields, state.temp_max),
            "version": state.version,
            **state.config_fields,
        }

    def metrics(range_value: str, n_points: int) -> JSON:
        points = [
            to_point(p, state.config_fields, state.temp_max)
            for p in make_points(range_value, state.scenario, n_points)
        ]
        return {"range": range_value, "points": points}

    class Handler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *args: object) -> None:
            log.debug("%s %s", self.address_string(), fmt % args)

        def _send(self, code: int, ctype: str, body: bytes | str) -> None:
            if isinstance(body, str):
                body = body.encode()
            try:
                self.send_response(code)
                self.send_header("Content-Type", ctype)
                self.send_header("Content-Length", str(len(body)))
                self.send_header("Access-Control-Allow-Origin", "*")
                self.end_headers()
                self.wfile.write(body)
            except (BrokenPipeError, ConnectionResetError):
                pass  # client went away mid-response

        def do_GET(self) -> None:
            try:
                self._route()
            except Exception:
                log.exception("error handling %s", self.path)
                self._send(500, "text/plain", "internal error")

        def _route(self) -> None:
            parsed = urlparse(self.path)
            path = parsed.path

            if path in ("/", "/index.html"):
                self._serve_dashboard()
            elif path == "/api/current":
                self._send(200, "application/json", dumps(current()))
            elif path == "/api/metrics":
                q = parse_qs(parsed.query)
                r = q.get("range", ["1d"])[0]
                n_points = clamp_points(q.get("points", [""])[0])
                self._send(200, "application/json", dumps(metrics(r, n_points)))
            elif path == "/stream":
                self._stream()
            else:
                self._serve_static(path)

        def _serve_dashboard(self) -> None:
            try:
                with open(DASHBOARD, "rb") as f:
                    self._send(200, "text/html; charset=utf-8", f.read())
            except FileNotFoundError:
                log.error("dashboard not found: %s", DASHBOARD)
                self._send(404, "text/plain", "dashboard not found")

        def _stream(self) -> None:
            try:
                self.send_response(200)
                self.send_header("Content-Type", "text/event-stream")
                self.send_header("Cache-Control", "no-cache")
                self.send_header("Access-Control-Allow-Origin", "*")
                self.end_headers()
                for _ in range(3):
                    self.wfile.write(f"data: {dumps(current())}\n\n".encode())
                    self.wfile.flush()
                    sleep(5)
            except (BrokenPipeError, ConnectionResetError):
                pass  # client closed the stream

        def _serve_static(self, path: str) -> None:
            # Serve static files from dashboard/: strict allowlist of known
            # filenames. Defence in depth: after looking up the filename in
            # the allowlist, realpath() resolves any symlink trickery and
            # startswith() confirms the result stays within DASHBOARD_DIR.
            # This is the path-injection sanitiser CodeQL recognises.
            filename = basename(path)
            if filename in STATIC_FILES:
                fpath = realpath(join(DASHBOARD_DIR, filename))
                base = realpath(DASHBOARD_DIR)
                if fpath.startswith(base + sep) and isfile(fpath):
                    with open(fpath, "rb") as f:
                        self._send(200, STATIC_FILES[filename], f.read())
                    return
            self._send(404, "text/plain", "not found")

    return Handler

File: tools/devserver/handler.js
/*
 * minimoni - zero-dependency system monitoring
 * Copyright (C) 2026 Javier Beaumont <javierbeaumont@users.noreply.github.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

/* HTTP request handler bound to an app state via makeHandler() (no module-global state). */

'use strict';

const fs = require('fs');
const path = require('path');

const { clampPointsParam, currentSnapshot, makePoints } = require('./mock-data');
const { toCurrent, toPoint } = require('./units');

const REPO_ROOT = path.join(__dirname, '..', '..');
const DASHBOARD_DIR = path.join(REPO_ROOT, 'dashboard');
const DASHBOARD = path.join(DASHBOARD_DIR, 'index.html');

const STATIC_FILES = {
  'app.js': 'text/javascript',
  'cards.js': 'text/javascript',
  'chart.js': 'text/javascript',
  'format.js': 'text/javascript',
  'hover.js': 'text/javascript',
  'style.css': 'text/css',
  'favicon.svg': 'image/svg+xml',
};

const NOOP_LOG = { debug() {}, info() {}, error() {} };

/* Flaky mode (--flaky): drop every other /stream connection so the SSE indicator
 * can be exercised (live -> reconnecting -> live). Deterministic by seq, off by
 * default. */
function streamShouldDrop(flaky, seq) {
  return flaky && seq % 2 === 1;
}

function send(res, code, ctype, body) {
  const buf = Buffer.isBuffer(body) ? body : Buffer.from(body);

  if (res.writableEnded)
    return;

  res.writeHead(code, {
    'Content-Type': ctype,
    'Content-Length': buf.length,
    'Access-Control-Allow-Origin': '*',
  });

  res.end(buf);
}

function makeHandler(state, log = NOOP_LOG) {
  let streamSeq = 0; /* per-server /stream connection counter */
  let staticBase = null; /* realpath(DASHBOARD_DIR): constant, resolved once on first static hit */

  /* converted mock metrics + real config fields */
  function current() {
    const raw = currentSnapshot(state.scenario);

    return {
      ...toCurrent(raw, state.configFields, state.tempCriticalFallback),
      ...state.configFields,
    };
  }

  function metrics(rangeValue, nPoints) {

    const points = makePoints(rangeValue, state.scenario, nPoints).map((p) =>
      toPoint(p, state.configFields, state.tempCriticalFallback)
    );

    return { range: rangeValue, points };
  }

  function serveDashboard(res) {
    let html;

    try {
      html = fs.readFileSync(DASHBOARD, 'utf8');
    } catch {
      log.error(`dashboard not found: ${DASHBOARD}`);

      send(res, 404, 'text/plain', 'dashboard not found');

      return;
    }

    /* Mirror bundle.sh: fill the {{VERSION}}/{{RELEASE}} placeholders (release
     * strips the git-describe "-N-gHASH" suffix so its link resolves). */
    const release = state.version.replace(/-\d+-g[0-9a-f]+$/, '');

    html = html.replace(/\{\{VERSION\}\}/g, state.version).replace(/\{\{RELEASE\}\}/g, release);

    send(res, 200, 'text/html; charset=utf-8', html);
  }

  function stream(req, res) {
    const seq = streamSeq++;

    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'Access-Control-Allow-Origin': '*',
    });

    if (streamShouldDrop(state.flaky, seq)) {
      log.debug(`flaky: dropping /stream connection ${seq}`);
      res.end(); /* close with no data: the browser EventSource errors */

      return;
    }

    let n = 0;
    let timer = null;

    const tick = () => {
      if (res.writableEnded)
        return;

      res.write(`data: ${JSON.stringify(current())}\n\n`);

      n += 1;
      timer = n < 3 ? setTimeout(tick, 5000) : setTimeout(() => res.end(), 5000);
    };

    req.on('close', () => {
      if (timer)
        clearTimeout(timer);
    });

    tick();
  }

  /* Serve dashboard/ static files via a strict filename allowlist; the realpath +
   * startsWith containment is defence-in-depth and the path-injection sanitiser
   * CodeQL recognises. */
  function serveStatic(res, urlPath) {
    const filename = path.basename(urlPath);
    const ctype = STATIC_FILES[filename];

    if (ctype) {
      try {
        const fpath = fs.realpathSync(path.join(DASHBOARD_DIR, filename));
        const base = (staticBase ??= fs.realpathSync(DASHBOARD_DIR));

        if (fpath.startsWith(base + path.sep) && fs.statSync(fpath).isFile()) {
          send(res, 200, ctype, fs.readFileSync(fpath));

          return;
        }
      } catch {
        /* fall through to 404 */
      }
    }

    send(res, 404, 'text/plain', 'not found');
  }

  function route(req, res) {
    const url = new URL(req.url, 'http://localhost');
    const p = url.pathname;

    if (p === '/' || p === '/index.html') {
      serveDashboard(res);
    } else if (p === '/api/current') {
      send(res, 200, 'application/json', JSON.stringify(current()));
    } else if (p === '/api/metrics') {
      const r = url.searchParams.get('range') || '1d';
      const nPoints = clampPointsParam(url.searchParams.get('points') || '');

      send(res, 200, 'application/json', JSON.stringify(metrics(r, nPoints)));
    } else if (p === '/stream') {
      stream(req, res);
    } else {
      serveStatic(res, p);
    }
  }

  return function handler(req, res) {
    /* Swallow client-disconnect errors (Node's BrokenPipe/ConnectionReset). */
    req.on('error', () => {});
    res.on('error', () => {});

    if (req.method !== 'GET') {
      send(res, 404, 'text/plain', 'not found');

      return;
    }

    log.debug(`${req.socket.remoteAddress} GET ${req.url}`);

    try {
      route(req, res);
    } catch (err) {
      log.error(`error handling ${req.url}: ${err.stack || err}`);

      send(res, 500, 'text/plain', 'internal error');
    }
  };
}

module.exports = { makeHandler, streamShouldDrop };
